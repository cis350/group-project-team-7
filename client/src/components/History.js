import React, { useState, useEffect } from "react";
import { useParams } from "react-router-dom";
import { Card, CardContent, Typography, Button } from "@mui/material";
import { useNavigate } from "react-router-dom";
import toast from 'react-hot-toast';

const serverUrl = process.env.REACT_APP_SERVER_URL;

const UserAnswers = () => {
    const { username } = useParams();
    const [userAnswers, setUserAnswers] = useState([]);
    const navigate = useNavigate();

    useEffect(() => {
        const fetchUserAnswers = async () => {
            try {
                const response = await fetch(`${serverUrl}/get_user_answers?username=${username}`, {
                    method: "GET",
                    credentials: 'include'
                });
                if (!response.ok) {
                    throw new Error("Failed to fetch user answers");
                }
                const data = await response.json();
                setUserAnswers(data);
            } catch (error) {
                console.error("Error fetching user answers:", error);
            }
        };

        fetchUserAnswers();
    }, [username]);

    const handleDeletePost = async (postId) => {
        try {
            const response = await fetch(`${serverUrl}/delete_one_answer?_id=${postId}`, {
                method: "DELETE",
                credentials: 'include',
            });
            if (!response.ok) {
                throw new Error("Failed to delete post");
            }
            toast.success("Deleted Answer!");
            setUserAnswers(userAnswers.filter((answer) => answer._id !== postId));
        } catch (error) {
            console.error("Error deleting post:", error);
        }
    };

    return (
        <div style={{ height: "100vh", padding: "5vh" }} className="w-100">
            <div className="text-center mb-4 text-4xl font-semibold">Your Post History:</div>
            <div className="grid grid-cols-3 gap-4">
                {userAnswers.map((answer, index) => (
                    <Card key={index} className="max-w-md mb-4">
                        <CardContent>
                            <Typography variant="h6" component="h2" className="mb-2">
                                Post {index + 1}
                            </Typography>
                            <Typography variant="body1" component="p" className="mb-2">
                                {answer.answer1}
                            </Typography>
                            <Typography variant="body1" component="p" className="mb-2">
                                {answer.answer2}
                            </Typography>
                            <Typography variant="body1" component="p" className="mb-2">
                                {answer.answer3}
                            </Typography>
                            <Typography variant="body1" component="p" className="mb-2">
                                {answer.answer4}
                            </Typography>
                            <Button onClick={() => handleDeletePost(answer._id)} color="error" variant="contained">
                                Delete
                            </Button>
                        </CardContent>
                    </Card>
                ))}
            </div>
            <div className="text-center">
                <Button
                    onClick={() => navigate("/form")}
                    variant="contained"
                    className="mt-4"
                >
                    Return to Form
                </Button>
            </div>
        </div>
    );
};

export default UserAnswers;
