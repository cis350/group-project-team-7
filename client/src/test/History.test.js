import React from 'react';
import { render, fireEvent, waitFor, screen } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import History from '../components/History';
import '@testing-library/jest-dom/extend-expect';
import { act } from 'react-dom/test-utils';
import { waitForElementToBeRemoved } from '@testing-library/dom';
import UserAnswers from '../components/History';
// Mock the navigation
const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
}));

describe('DataVisualization', () => {
  let consoleSpy;
  // Setup mock fetch in the global scope
  beforeEach(() => {
    global.fetch = jest.fn(() =>
        Promise.resolve({
            // ok: false, // Set ok property to false for unsuccessful response
            // status: 401, // Unauthorized response for the login attempt
            // json: () => Promise.resolve({ message: "Unauthorized" }),
            // text: () => Promise.resolve("No current user")
            ok: true,
            json: () => Promise.resolve([{ answer1: "10", answer2: "20", username: "user1" }]),
            text: () => Promise.resolve("Current user information")
        })
    );
    consoleSpy = jest.spyOn(console, 'error').mockImplementation(() => {});
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders the component and matches the snapshot', async () => {
    // Setup fetch to return some data
    global.fetch.mockResolvedValueOnce({
      ok: true,
      json: async () => [{ answer1: "10", answer2: "20", answer3: "30", username: "user1" }],
    });

    const { asFragment } = render(<Router><UserAnswers /></Router>);
    await waitFor(() => {
        expect(screen.getByText("Your Post History:")).toBeInTheDocument();
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it('fetches data and displays it correctly', async () => {
    // Return mock data for the fetch call
    global.fetch.mockResolvedValueOnce({
      ok: true,
      json: async () => [{ answer1: "Yes", answer2: "No", answer3: "Maybe", username: "testUser" }],
    });

    render(<Router><UserAnswers /></Router>);

    await waitFor(() => {
        expect(screen.getByText("Your Post History:")).toBeInTheDocument();
    });
  });

  it('handles fetch errors gracefully', async () => {
    global.fetch.mockRejectedValueOnce(new Error('Failed to fetch data'));
    const { getByText } = render(
      <Router>
        <UserAnswers />
      </Router>
    );
  });

});
