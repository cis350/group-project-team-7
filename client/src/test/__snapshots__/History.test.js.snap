// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`History renders the component and matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="w-100"
    style="height: 100vh; padding: 5vh;"
  >
    <div
      class="text-center mb-4 text-4xl font-semibold"
    >
      Your Post History:
    </div>
    <div
      class="grid grid-cols-3 gap-4"
    />
    <div
      class="text-center"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium mt-4 css-sghohy-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Return to Form
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
